setup_mysql.sh

```diff
@@ -0,0 +1,230 @@
+  # -c flag prefaces a command to be executed before any in the script
+  # Use HERE document vs. as a straight multiline string - to avoid quoting issues
+  echo
+  echo "GETTING assigned root password from file..."
+  MYSQL_ROOT_PWD=$(<${MYSQL_PW_FILE})
+  echo "new root password: ${MYSQL_ROOT_PWD}"
+  echo
+
+SECURE_MYSQL=$(expect -f '-' <<HERE
+set timeout 10
+spawn mysql_secure_installation
+
+expect "Enter password for user root:"
+send "${MYSQL_TEMP_PWD}\r"
+
+expect {
+  "Change the password for root ? \(Press y|Y for Yes, any other key for No) :" {
+    send "y\r"
+  }
+  "The existing password for the user account root has expired. Please set a new password." {
+  }
+}
+expect "New password:"
+send "${MYSQL_ROOT_PWD}\r"
+
+expect "Re-enter new password:"
+send "${MYSQL_ROOT_PWD}\r"
+
+expect {
+  "Do you wish to continue with the password provided?\(Press y|Y for Yes, any other key for No) :" {
+    send "y\r"
+  }
+  "Change the password for root ? \(\(Press y|Y for Yes, any other key for No) :" {
+    send "n\r"
+  }
+}
+
+expect "Remove anonymous users? \(Press y|Y for Yes, any other key for No) :"
+send "y\r"
+
+expect "Disallow root login remotely? \(Press y|Y for Yes, any other key for No) :"
+send "y\r"
+
+expect "Remove test database and access to it? \(Press y|Y for Yes, any other key for No) :"
+send "y\r"
+
+expect "Reload privilege tables now? \(Press y|Y for Yes, any other key for No) :"
+send "y\r"
+expect eof
+HERE
+)
+echo "$SECURE_MYSQL"
+echo "RAN mysql_secure_installation & root password has been assigned for MySQL."
+echo
+}
+
+# Check if MySQL IS installed, i.e., you already set up MySQL yet & already
+# changed temporary password - so NO need to install MySQL.
+# Next: Check if mysql_secure_installation has been run.
+# -v means tell if pattern not found
+if [ "$(sudo yum repolist enabled | grep "mysql.*-community.*")" != '' ] && command -v mysqld; then
+  echo "MySQL package repos AND MySQL server are already installed."
+  echo
+
+  # Case 1: CHECK if assignedroot password works with MySQL.
+  # If it root password exists, exit with 0 & skip rest of code since it means
+  # MySQL is installed AND mysql_secure_installation has already been run.
+  # Case 2: else - Only the temporary password exists so far which means MySQL
+  # is installed BUT mysql_secure_installation has NOT been run yet, so run it.
+  echo "TEST if root password runs MySQL, i.e., mysql_secure_installation was already run."
+  echo
+  if expect -f '-' <<HERE
+  set timeout 5
+  spawn mysql -u root -p --execute "quit"
+  expect "Enter password:"
+  send "$MYSQL_ROOT_PWD\r"
+  expect eof
+  exit [lindex [wait] 3]
+HERE
+  then
+    echo
+    echo "ROOT password assigned in mysql_pw.txt WORKED! MySQL is installed & mysql_secure_installation was run."
+    exit
+  else
+    echo
+    echo "NO assigned ROOT password yet. RUNNING mysql_secure_installation..."
+    echo
+    run_mysql_secure_installation
+  fi
```

Comment:
I think you don't need this code from lines 99 through 123, because you already handle running `mysql_secure_installation` later in the script.

Reply:
I think that code is still a valuable check to exit early if an assigned root password exists, i.e., MYSQL_ROOT_PWD. So I moved it to after Line 146-147:  
```
sudo systemctl start mysqld.service
sudo systemctl status mysqld
```
So it is now before Line 169:  
`# CHECK ODD CASE`
and before Line 188:    
`# CHECK if temporary password can be used to RUN MySQL.`

And I removed the else.

This change also took care of an error seen before that I couldn't re-create:
...
spawn mysql_secure_installation

Securing the MySQL server deployment.

Enter password for user root:
Error: Can't connect to local MySQL server through socket '/var/lib/mysql/mysql.sock' (2)
expect: spawn id exp6 not open
    while executing
"expect "New password:""

Reply:
@kcharter I created a new clean VM, copied over the required files & re-tested the script in various states of removal with these commands:  
```
sudo yum erase mysql  # Removes the client.
sudo rm -rf /var/lib/mysql  # Deletes the database files.*
sudo yum erase mysql-community-server  # Deletes server config files.
sudo rm /var/log/mysqld.log
```

Reply:
Yep, looking at it this afternoon.